### ISSUE TITLE
Descriptors cannot not be created directly.

### ISSUE THREAD

**Issue by Veldkornet** [snicker/zwift_hass#40]
My Zwift sensors no longer work since updating to 3.3.1, I get the following error during startup. 

```
2023-05-05 22:58:43.933 ERROR (MainThread) [custom_components.zwift.sensor] Could not create Zwift sensor named 'Zwift'!
Traceback (most recent call last):
  File "/config/custom_components/zwift/sensor.py", line 124, in async_setup_platform
    await zwift_data._connect()
  File "/config/custom_components/zwift/sensor.py", line 343, in _connect
    from zwift import Client as ZwiftClient
  File "/usr/local/lib/python3.10/site-packages/zwift/__init__.py", line 4, in <module>
    from .client import Client
  File "/usr/local/lib/python3.10/site-packages/zwift/client.py", line 5, in <module>
    from .world import World
  File "/usr/local/lib/python3.10/site-packages/zwift/world.py", line 2, in <module>
    from . import zwift_messages_pb2
  File "/usr/local/lib/python3.10/site-packages/zwift/zwift_messages_pb2.py", line 35, in <module>
    _descriptor.FieldDescriptor(
  File "/usr/local/lib/python3.10/site-packages/google/protobuf/descriptor.py", line 561, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors cannot not be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

More information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates
```

**Comment by danka621** [comment]
I have the same problem

**Comment by d4rthpaul** [comment]
I also have the same problem

**Comment by clyra** [comment]
Hi all,

I also had the same problem and version 3.3.1 didnt worked. It should... but it seems that zwift-client package just ignores the PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION environment variable. As https://github.com/jsmits/zwift-client/issues/377 suggests, the _pb2 package needs to be regenerated, so I did it:

````
# pip install grpcio-tools
# cd /usr/local/lib/python3.11/site-packages/zwift
# wget https://raw.githubusercontent.com/jsmits/zwift-client/master/zwift/zwift_messages.proto  
# python -m grpc_tools.protoc -I .  --python_out=. --pyi_out=. --grpc_python_out=. zwift_messages.proto
````

**Comment by Veldkornet** [comment]
You did that from inside the homeassistant container?
I get a bunch of dependency errors with the pip install

**Comment by clyra** [comment]
yes, I did from inside the homeassistant container. I guess that since HA may install extra packages based on your custom integrations, mine already had all the needed... Based on your screenshot, it seems that your container doesnt have the c++ compiles.  Try to "apk update" and "apk add g++". But, it still looks strange, because i didnt had to install g++ ! which version of HA are you running?

**Comment by Veldkornet** [comment]
Well I’m on 2023.7.3 at the moment. 

I did get g++ installed. But then the pip install of grpcio-tools maxed out the CPU of my Pi for like 3 hours and then eventually failed with some massive error stack that I couldn’t even get to the beginning of.

**Comment by clyra** [comment]
hm...maybe there's no compiled package for that architecture. Anyway... you may try to unzip the following file on /usr/local/lib/python3.11/site-packages/zwift. 

[zwift.zip](https://github.com/snicker/zwift_hass/files/12214213/zwift.zip)

**Comment by clyra** [comment]
hm... it seems that my previous solution (the "pip install" one) is not good as the pip changes another package needed by HA. Just unpacking the zip files works for me.

**Comment by danka621** [comment]
> hm...maybe there's no compiled package for that architecture. Anyway... you may try to unzip the following file on /usr/local/lib/python3.11/site-packages/zwift.
> 
> [zwift.zip](https://github.com/snicker/zwift_hass/files/12214213/zwift.zip)

How did you get to issue commands inside a container?

I dont know how to start a shell in my installation. The suggested method I find on forums involve choosing a SSH terminal from Add-on store, but there is no add-on store in my home assistant.

**Comment by Veldkornet** [comment]
> > hm...maybe there's no compiled package for that architecture. Anyway... you may try to unzip the following file on /usr/local/lib/python3.11/site-packages/zwift.
> 
> > 
> 
> > [zwift.zip](https://github.com/snicker/zwift_hass/files/12214213/zwift.zip)
> 
> 
> 
> How did you get to issue commands inside a container?
> 
> 
> 
> I dont know how to start a shell in my installation. The suggested method I find on forums involve choosing a SSH terminal from Add-on store, but there is no add-on store in my home assistant.

You can also just use the Portainer add-on to open a console on the container

**Comment by danka621** [comment]
Portainer? Does that also require the add-on store?

**Comment by Veldkornet** [comment]
> Portainer? Does that also require the add-on store?

It is in the add-on store. But no, you can use it anywhere. It's just a container itself.

**Comment by danka621** [comment]
So if I dont have add-on store, I cant install portainer, correct?

**Comment by Veldkornet** [comment]
> So if I dont have add-on store, I cant install portainer, correct?

Correct

**Comment by Veldkornet** [comment]
> hm...maybe there's no compiled package for that architecture. Anyway... you may try to unzip the following file on /usr/local/lib/python3.11/site-packages/zwift. 
> 
> 
> 
> [zwift.zip](https://github.com/snicker/zwift_hass/files/12214213/zwift.zip)
> 
> 

I'll give this a try

**Comment by danka621** [comment]
Thanks @clyra! I got it working on my HA container installation.
I could copy the files in place via shell commands in configuration.yaml:

**shell_command:**
__**backup_old_zwiftfile**: cp /usr/local/lib/python3.11/site-packages/zwift/zwift_messages_pb2.py /config/original_zwift_messages_pb2.py
__**copynewfile:** cp /config/clyras_files/zwift_messages_pb2.py /usr/local/lib/python3.11/site-packages/zwift/zwift_messages_pb2.py 
 __**copynewfile2:** cp /config/clyras_files/zwift_messages_pb2_grpc.py /usr/local/lib/python3.11/site-packages/zwift/zwift_messages_pb2_grpc.py
    

There is also the possibilty to navigate the docker image according to this post: https://stackoverflow.com/a/74515438

**Comment by Bikechimp** [comment]
Hi there,
i have a fresh hass system installed and also managed to get the hacs store installed. I now have installed the zwift sensor package, newest version.
Task: i know have to install a sensor (i am completely new on this and need help how to do this)
sensor:
  - platform: zwift
    username: !secret my_zwift_username
    password: !secret my_zwift_password
    players:
      - !secret my_friends_zwift_player_id
Is there a video or manual what to do, to get the sensors going..
i would be greatlyx happy, just to see the sensor/entities on my system..the rest will be done in the upcoming months..
thanks for your help

**Comment by Bikechimp** [comment]
I forgot to mention, i have a x86 thin client with hass installed only, no proxmox, no virtual machine...just a clean build which runs zigbee and other stuff succesfully

**Comment by snicker** [comment]
Closing as duplicate of #37 and re-opening that issue. Please direct discussion there.

**Comment by snicker** [comment]
fixed in 3.3.2 (https://github.com/snicker/zwift_hass/releases/tag/v3.3.2)
